### ISSUE TITLE
Execute "dmenu" brings a repeted key in a loop

### ISSUE THREAD

**Issue by spectrum70** [labwc/labwc#176]
Hi,
thanks for the great job.
I am testing git master, 2ce961a0bd2d92e091f993d0f63cd01bf06158c7
I have

`
 <keybind key="W-d">
     <action name="Execute">
        <command>dmenu_run -f -fn "Conthrax-9" -nb "#2020ff" -nf "#ffff00"</command>
     </action>
  </keybind>
`
On the first execution of W-d , after labwc just launched, all is ok, dmenu works properly.
Second time, even if i just press "W-d" and release it immediatley, a "d" is issued in dmenu, continuously (i see "ddddddd"), and i have to stop it by pressing a key and delete them all.

If i launch "dmenu_run" from console, all works properly.

**Comment by johanmalm** [comment]
Thanks for reporting. Looks like this issue has been unmasked by commit:

https://github.com/labwc/labwc/commit/7e57b7fcf51e256f1c786e905de6dfab0411463e

Before that commit, the release event for the “d” would have been passed to dmenu this cancelling the repeat.

Looks like we ought to disarm repeat before actioning keybinds:

https://github.com/swaywm/sway/blob/4e77bc293515ea2f00ebbc4395cb0eb4446d0195/sway/input/keyboard.c#L487

**Comment by spectrum70** [comment]
Thanks.
Would like to contribute :) Is it ok if i try a patch for this ?

**Comment by johanmalm** [comment]
Great. Yes, go for it.

**Comment by johanmalm** [comment]
> Looks like we ought to disarm repeat before actioning keybinds:
> 
> https://github.com/swaywm/sway/blob/4e77bc293515ea2f00ebbc4395cb0eb4446d0195/sway/input/keyboard.c#L487

Apologies. I think the above is a red herring. Not sure what the answer is.
Any thoughts from anyone???

**Comment by spectrum70** [comment]
Hi Johan and all, 

i have 2 fixes, one working well, but using a usleep, that is not likely the right solution. The other using a timer, but losing first "d" on the way (need to press 'd' 2 times calling dmenu program starting for 'd'). Let me rearrange the code a bit.

Common
`
diff --git a/src/action.c b/src/action.c
index 926db06..0c00705 100644
--- a/src/action.c
+++ b/src/action.c
@@ -42,7 +42,9 @@ action(struct view *activator, struct server *server, const char *action, const
                buf_init(&cmd);
                buf_add(&cmd, command);
                buf_expand_shell_variables(&cmd);
+               keyboard_repeat_disarm(&server->seat);
                spawn_async_no_shell(cmd.buf);
+               keyboard_repeat_rearm(&server->seat);
                free(cmd.buf);
        } else if (!strcasecmp(action, "Exit")) {
                wl_display_terminate(server->wl_display);
`

Solution 1

`
void
keyboard_repeat_disarm(struct seat *seat)
{
	struct wlr_keyboard *kb = &seat->keyboard_group->keyboard;

	wlr_keyboard_set_repeat_info(kb, 0, 0);
}

void
keyboard_repeat_rearm(struct seat *seat)
{
	struct wlr_keyboard *kb = &seat->keyboard_group->keyboard;

	usleep(120000);
	wlr_keyboard_set_repeat_info(kb, rc.repeat_rate, rc.repeat_delay);
}

Solution 2
`void
keyboard_repeat_disarm(struct seat *seat)
{
	struct wlr_keyboard *kb = &seat->keyboard_group->keyboard;
	struct wl_display *display;
	struct wl_event_loop *loop;
	struct wl_event_source *rearm_timer;

	wlr_keyboard_set_repeat_info(kb, 0, 0);

	display = seat->server->xwayland->wl_display;
        loop = wl_display_get_event_loop(display);

	rearm_timer = wl_event_loop_add_timer(
			loop, keybpard_rearm_timeout, seat);
	wl_event_source_timer_update(rearm_timer, rc.repeat_delay + 10);
}

void
keyboard_repeat_rearm(struct seat *seat)
{;
}
`

I am quite new to wayland,  so both solutions are likely rudimental and improper.
Will be out until jan 10, then can come back on this. Solution 1 works fine for me as a worklaround. But, if in the meantime youo want to fix this properly, no problem.

**Comment by johanmalm** [comment]
I'm not sure if we're missing something here. `sway` doesn't seem to have to cancel/reset the repeat.
I've pushed a fix for the time being based on your code - but without the timer/usleep.
The only oddity is that we lose the first "d" as you described above for solution 2. I can't quite workout why.
I've tried with `rofi` and `leafpad`, but don't see the same behavior with them.

**Comment by johanmalm** [comment]
I tried some `xkb_state_update_key` but with no success.

```
diff --git a/include/key-state.h b/include/key-state.h
index e5f7291..addc492 100644
--- a/include/key-state.h
+++ b/include/key-state.h
@@ -2,11 +2,14 @@
 #ifndef __LABWC_KEY_STATE_H
 #define __LABWC_KEY_STATE_H
 
+struct wlr_keyboard;
+
 void key_state_set_pressed(uint32_t keycode, bool ispressed);
 void key_state_store_pressed_keys_as_bound(void);
 bool key_state_corresponding_press_event_was_bound(uint32_t keycode);
 
 /* returns numbers of keys still pressed in a consumed key combination */
 int key_state_bound_key_remove(uint32_t keycode);
+void key_state_clear_xkb_state(struct wlr_keyboard *kb);
 
 #endif /* __LABWC_KEY_STATE_H */
diff --git a/src/key-state.c b/src/key-state.c
index d2be546..cc54414 100644
--- a/src/key-state.c
+++ b/src/key-state.c
@@ -2,6 +2,7 @@
 #include <stdbool.h>
 #include <stdint.h>
 #include <string.h>
+#include <wlr/types/wlr_keyboard.h>
 
 #define MAX_PRESSED_KEYS (16)
 
@@ -69,3 +70,11 @@ key_state_bound_key_remove(uint32_t keycode)
 	remove_key(&bound, keycode);
 	return bound.nr_keys;
 }
+
+void
+key_state_clear_xkb_state(struct wlr_keyboard *kb)
+{
+	for (int i = 0; i < pressed.nr_keys; ++i) {
+		xkb_state_update_key(kb->xkb_state, pressed.keys[i], XKB_KEY_UP);
+	}
+}
diff --git a/src/keyboard.c b/src/keyboard.c
index c10e3c9..c6105c3 100644
--- a/src/keyboard.c
+++ b/src/keyboard.c
@@ -67,6 +67,7 @@ handle_keybinding(struct server *server, uint32_t modifiers, xkb_keysym_t sym)
 		for (size_t i = 0; i < keybind->keysyms_len; i++) {
 			if (xkb_keysym_to_lower(sym) == keybind->keysyms[i]) {
 				wlr_keyboard_set_repeat_info(kb, 0, 0);
+				key_state_clear_xkb_state(kb);
 				action(NULL, server, keybind->action,
 				       keybind->command, 0);
 				return true;
```

**Comment by Consolatis** [comment]
I experience a related issue in that at some random point repeating is disabled completely.
When restarting labwc (nested under X11) the issue is gone and repeating works as expected until it hits again.
So far I failed to reproduce it but detected that when using a keybind like `W-Tab` the application with keyboard focus gets a press event for the modifier (`Super` in this case) but never a release.

<details><summary>Maybe this helps for future debugging</summary>

```
#!/usr/bin/env python3

import gi

gi.require_version("Gtk", "3.0")
gi.require_version("Gdk", "3.0")
from gi.repository import Gtk, Gdk

class Window(Gtk.Window):
	def __init__(self):
		super().__init__(title="Input test")
		self.connect('key_press_event', self.on_key_press)
		self.connect('key_release_event', self.on_key_release)
		self.connect('destroy', lambda window : Gtk.main_quit())
		self.fmt = "  {:{pad}>6s}  {:{pad}>7s}  {:{pad}>6s}  {:{pad}>6s}  {:{pad}>7s}  {:{pad}<20s}"
		self.fmt = self.fmt.format

		print()
		self.log("Source", "Action", "Code", "KeySym", "Key", "States")
		self.log("", "", "", "", "", "", pad='-')

	def on_key_press(self, window, event):
		self.log('Key', "press", event.hardware_keycode, event.keyval, event.string, event.state)

	def on_key_release(self, window, event):
		self.log('Key', "release", event.hardware_keycode, event.keyval, event.string, event.state)

	def log(self, source, action, keycode, keysym, content, states, pad=''):
		if content == '\t':
			content = '[tab]'
		elif content == ' ':
			content = '[space]'
		print(self.fmt(source, action, str(keycode), str(keysym), content, str(states), pad=pad))

if __name__ == '__main__':
	window = Window()
	window.show_all()
	Gtk.main()
```
</details>

**Comment by spectrum70** [comment]
Hi, thanks for the fix, i readded the sleep workaround, maybe better than nothing now

From cf745cc9512a2d7960ab7c955a13362f8d07c517 Mon Sep 17 00:00:00 2001
From: Angelo Dureghello <angelo@kernel-space.org>
Date: Mon, 10 Jan 2022 00:51:04 +0100
Subject: [PATCH] fix first key eaten after keybind action

I.e., on a W-d action executing dmenu, first 'd' pressed on dmenu
is  eaten and a second 'd' should be pressed. It is actually not
clear who is eating the 'd', anyway, fixing this for now
with a dirty (usleep) workaround.

```
Signed-off-by: Angelo Dureghello <angelo@kernel-space.org>
---
 src/keyboard.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/src/keyboard.c b/src/keyboard.c
index 9b1bf84..ffae88e 100644
--- a/src/keyboard.c
+++ b/src/keyboard.c
@@ -4,6 +4,9 @@
 #include "action.h"
 #include "key-state.h"
 #include "labwc.h"
+#include <time.h>
+
+#define DELAY_NOREPEAT         140000

 static void
 change_vt(struct server *server, unsigned int vt)
@@ -67,8 +70,15 @@ handle_keybinding(struct server *server, uint32_t modifiers, xkb_keysym_t sym)
                }
                for (size_t i = 0; i < keybind->keysyms_len; i++) {
                        if (xkb_keysym_to_lower(sym) == keybind->keysyms[i]) {
+                               struct timespec delay = {0, DELAY_NOREPEAT * 1000};
+
                                wlr_keyboard_set_repeat_info(kb, 0, 0);
                                action(NULL, server, &keybind->actions, 0);
+                               /*
+                                * Workaround: a brief delay after the action
+                                * allows to avoid the first key to be eaten.
+                                */
+                               nanosleep(&delay, 0);
                                return true;
                        }
                }
--
2.34.1
```

**Comment by Consolatis** [comment]
Alright, spent some more time on this bug and started by reverting e62bb51bfb33ee520e800cf98553d766824fe9cf + adding a bunch of debug logs based on this statement:
> I'm not sure if we're missing something here. sway doesn't seem to have to cancel/reset the repeat.

What I can say so far that it doesn't make any sense yet.
The repeat timer only fires for X11 clients and only after the second time being used (nothing new here).
We are not sending \*any\* keys other than the first press of the modifier before executing the binding (which is a separate thing to solve at some point, hacked a really bad fix into the test branch but that wasn't the actual issue).

Current debug branch can be found [here](https://github.com/Consolatis/labwc/tree/fix/key_repeat)

**Comment by Consolatis** [comment]
Should be fixed by #438
